<a href="https://colab.research.google.com/github/samanthajmichael/dsba6211-summer2024/blob/main/notebooks/dsba6211_summer2024_lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 2 Instructions

For these problems, you'll use [scikit-learn](https://scikit-learn.org/stable/). Feel free to use the documentation as reference.

There are two parts to do:
1. code fill-in the blank
2. markdown written comments

Code fill-in the blanks are code snippets with:
```
# fill in the blank:
```

For markdown written comments, replace the text with:

`<!-- Write in markdown your answer below -->`

As normal, please complete this Colab notebook, save it to your DSBA6211 GitHub repo, and then submit that URL to that notebook on Canvas.


# Problem 1

A pharmaceutical company is conducting a market research project to gauge consumer interest in buying a **cold sore healthcare product**. To do so, they've collected data from a sample of **1,023 consumers**.

As part of the study, each participant was asked to evaluate **35 key factors** that influence their decision to purchase a cold sore treatment, rating each factor on a **10-point scale** to indicate its level of importance.

Additionally, they were asked to rate their overall interest in buying the product (y). This is the variable we're most interested in analyzing.

In [1]:
import pandas as pd

data_url = "https://github.com/DS-100/textbook/raw/master/content/datasets/market-analysis.csv"

df = pd.read_csv(data_url)

df.head(n=5)

y  x1  x2  x3  x4  x5  x6  x7  x8  x9  ...  x26  x27  x28  x29  x30  x31  \
0  7   8   8   8   8   8   8   8   8   8  ...    8    8    8    7    8    8   
1  9   9   9   9   9   9   7   9   9   9  ...    9    9    9    9    9    9   
2  5   6   6   5   5   8   5   7   5   3  ...    8    8    6    7    7    5   
3  6   7   5   8   6   8   6   7   4   6  ...    7    7    6    6    7    8   
4  5   4   6   4   4   3   4   5   4   4  ...    5    5    5    5    5    4   

   x32  x33  x34  x35  
0    8    5    8    8  
1    9    9    3    9  
2    8    5    4    4  
3    6    3    6    8  
4    6    4    5    6  

[5 rows x 36 columns]

Here's a table that provides a description for each of the 35 variables and its correlation with `y`.

|  | Corr | Description |  | Corr | Description |
| --- | --- | --------- | --- | --- | --------- |
| x1  | 0.70 | provides soothing relief | x19 | 0.54 | has a non-messy application |
| x2  | 0.58 | moisturizes cold sore blister | x20 | 0.70 | good for any stage of a cold |
| x3  | 0.69 | provides long-lasting relief | x21 | 0.49 | easy to apply/take |
| x4  | 0.70 | provides fast-acting relief | x22 | 0.52 | package keeps from contamination |
| x5 | 0.72 | shortens duration of a cold | x23 | 0.57 | easy to dispense a right amount |
| x6  | 0.68 | stops the virus from spreading | x24 | 0.63 | worth the price it costs |
| x7 | 0.67| dries up cold sore | x25 | 0.57 | recommended most by pharamacists |
| x8 | 0.72 | heals fast | x26 | 0.54 | recommended by doctors |
| x9 | 0.72 | penetrates deep | x27 | 0.54 | FDA approved |
| x10 | 0.65 | relieves pain | x28 | 0.64 | a brand I trust |
| x11 |0.61 | prevents cold | x29 | 0.60 | clinically proven |
| x12 | 0.73 | prevents from getting worse | x30 | 0.68 | a brand I would recommend |
| x13 | 0.57 | medicated | x31 | 0.74 | an effective treatment |
| x14 | 0.61 | prescription strength | x32  |0.37 | portable |
| x15 | 0.63 | repairs damaged skin | x33 | 0.37 | discreet packaging |
| x16 | 0.67 | blocks virus from spreading | x34 | 0.55 | helps conceal cold sores |
| x17 | 0.42 | contains SPF | x35 | 0.63 | absorbs quickly |
| x18 | 0.57 | non-irritating | | | |

You can replicate the correlations like this:

In [2]:
round(df[['x1','x2','y']].corr(),2)

x1    x2     y
x1  1.00  0.69  0.70
x2  0.69  1.00  0.58
y   0.70  0.58  1.00

For this question, we'll convert the y variable to a binary to help us frame the problem as a binary classification.

In [3]:
# if df["y"] > 7 then 1
y = df["y"].apply(lambda x: 1 if x > 7 else 0)
X = df.drop(columns=["y"])

## Fill in the blanks

Create a `train_test_split` on your `X` and `y` data. Set the `test_size=200` and set the `random_state` to `42`.

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 200, random_state=42)

Create scikit-learn pipeline with StandardScalar and LogisticRegression. Use 42 for the `random_state`.



In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

pipe = Pipeline([
    ('scale', StandardScaler()),
    ('model', LogisticRegression(random_state=42))
])

pipe.fit(X_train, y_train)

Pipeline(steps=[('scale', StandardScaler()),
                ('model', LogisticRegression(random_state=42))])

Measure test (out-of-sample) accuracy using `accuracy_score`.

In [6]:
from sklearn.metrics import accuracy_score
accuracy_score(pipe.predict(X_test), y_test)

0.79

Let's now explore what happens if we modify regularization by modifying hyperparameters.

For this, I recommend referencing [this tutorial on scikit-learn hyperparameters](https://inria.github.io/scikit-learn-mooc/python_scripts/parameter_tuning_manual.html)

Show all of the parameters names (hint: you'll need to find the method on `pipe` that displays all of `pipe`'s parameters).

In [7]:
pipe.get_params()

{'memory': None,
 'steps': [('scale', StandardScaler()),
  ('model', LogisticRegression(random_state=42))],
 'verbose': False,
 'scale': StandardScaler(),
 'model': LogisticRegression(random_state=42),
 'scale__copy': True,
 'scale__with_mean': True,
 'scale__with_std': True,
 'model__C': 1.0,
 'model__class_weight': None,
 'model__dual': False,
 'model__fit_intercept': True,
 'model__intercept_scaling': 1,
 'model__l1_ratio': None,
 'model__max_iter': 100,
 'model__multi_class': 'auto',
 'model__n_jobs': None,
 'model__penalty': 'l2',
 'model__random_state': 42,
 'model__solver': 'lbfgs',
 'model__tol': 0.0001,
 'model__verbose': 0,
 'model__warm_start': False}

With the name of the C hyperparameter from the previous step, use the `set_params` to train a model with a C value of `100`.

In [8]:
pipe.set_params(model__C=100)

Pipeline(steps=[('scale', StandardScaler()),
                ('model', LogisticRegression(C=100, random_state=42))])

Now use `cross_validate` (I've provided the import below) to do cross-validation to test these different values of C: `[1e-3, 1e-2, 1e-1, 1, 10, 100]`

You will need to provide the mean and +/- 1 std dev for each value of C. Be sure to carefully select the right datasets to use.

(Hint: see the tutorial on scikit-learn hyperparameters provided above)

In [9]:
from sklearn.model_selection import cross_validate
for C in [1e-3, 1e-2, 1e-1, 1, 10, 100]:
  pipe.set_params(model__C=C)
  cv_results = cross_validate(pipe, X_train, y_train)
  scores = cv_results["test_score"]
  print(
        f"Accuracy score via cross-validation with C={C}:\n"
        f"{scores.mean():.3f} ± {scores.std():.3f}"
    )

Accuracy score via cross-validation with C=0.001:
0.818 ± 0.026
Accuracy score via cross-validation with C=0.01:
0.823 ± 0.021
Accuracy score via cross-validation with C=0.1:
0.812 ± 0.025
Accuracy score via cross-validation with C=1:
0.804 ± 0.022
Accuracy score via cross-validation with C=10:
0.803 ± 0.023
Accuracy score via cross-validation with C=100:
0.803 ± 0.023


Based on your cross-validation, which hyperparameter would you suggest to use? How large of an impact do the different possible C values have?

#Based on the cross validation, the most appropriate hyperparameter is 0.01 which yielded the highest accuracy at 82.3% and the lowest standard deviation.

#Smaller values of C have a larger impact on regularization of the model suggesting that moderate regularization is optimal for this data set. As the value of C increases past 0.01, the accuracy score decreases.

# Problem 2: problems in pipelines

Let's assume that you have run this model.

In [39]:
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

random_state = 42
X, y = make_classification(random_state=random_state, n_features=10)
print(X.shape, y.shape)

(100, 10) (100,)


In [40]:
print(X[0:5],y[0:5])

[[-1.14052601  1.35970566  0.86199147  0.84609208  0.60600995 -1.55662917
   1.75479418  1.69645637 -1.28042935 -2.08192941]
 [-0.07873421 -1.32933233  0.6273745  -1.19300559 -0.77300978  0.0976761
   0.49799829  0.95927083  0.02451017  1.45114361]
 [ 0.80742726  0.73019848 -1.28568005  0.88948365 -1.8048821  -0.76325916
   0.04808495 -0.90431663 -1.62754244  0.2597225 ]
 [ 0.58846525 -0.3751207  -0.57500215 -0.14951801 -0.56372455  0.41293145
   0.24368721 -0.50694318 -0.8222204   0.24496657]
 [ 1.63631177 -1.64060704 -1.36045573 -0.94116325 -1.43014138  1.6324113
   0.13074058 -1.43586215 -0.44004449  1.44127329]] [1 1 0 0 0]


We'll partition the data into a train-test split with 30% in the test dataset.

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=random_state)

Let's now assume that you were told to standardize your X variable like so:

In [13]:
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_transformed = scaler.fit_transform(X_train)
model = LogisticRegression(C=1).fit(X_train_transformed, y_train)
accuracy_score(y_test, model.predict(X_test))

0.9666666666666667

## Question

What is a problem with the above workflow?

*Hint: go step by step, and view all of the outputs and inputs*

# The fit_transform() method is used on the training data (X_train). This method fits the scaler to the training data and transforms it. However, the fitted scaler should only be used to transform the test data (X_test), not to fit it.

Write in the code snippet below a corrected version of your model.

*Hint: rewrite to fix the problem (2 lines of code)*

In [14]:
X_test_transformed = scaler.transform(X_test)
accuracy_score(y_test, model.predict(X_test_transformed))

0.9333333333333333

# Problem 3: Building pipelines

For this problem, we'll use `skrub`, a helpful `scikit-learn` extension package.


In [15]:
!pip install skrub

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 142.6/142.6 kB 2.0 MB/s eta 0:00:00


In [16]:
from sklearn.datasets import fetch_openml

X, y = fetch_openml("titanic", version=1, as_frame=True, return_X_y=True)

X.head(3)

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


pclass                            name     sex      age  sibsp  parch  \
0     1.0   Allen, Miss. Elisabeth Walton  female  29.0000    0.0    0.0   
1     1.0  Allison, Master. Hudson Trevor    male   0.9167    1.0    2.0   
2     1.0    Allison, Miss. Helen Loraine  female   2.0000    1.0    2.0   

   ticket      fare    cabin embarked  boat  body  \
0   24160  211.3375       B5        S     2  None   
1  113781  151.5500  C22 C26        S    11   NaN   
2  113781  151.5500  C22 C26        S  None   NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON

Create a train vs test partition, holding out 20% of the data.

In [17]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=42)

Use the code below to create a simple `scikit-learn` pipeline with only `pclass` and `sex` encoded as a One Hot Encoding:

In [18]:
from sklearn.pipeline import Pipeline, make_pipeline,FeatureUnion, make_union
from sklearn.preprocessing import OneHotEncoder
from skrub import SelectCols

pclass_sex_pipe = make_union(
    Pipeline([
    ('pclass_col', SelectCols('pclass')),
    ('pclass_enc', OneHotEncoder(sparse_output=False))
  ]),
    Pipeline([
    ('sex_col', SelectCols('sex')),
    ('sex_enc', OneHotEncoder(sparse_output=False))
  ])
)
pclass_sex_pipe

FeatureUnion(transformer_list=[('pipeline-1',
                                Pipeline(steps=[('pclass_col',
                                                 SelectCols(cols='pclass')),
                                                ('pclass_enc',
                                                 OneHotEncoder(sparse_output=False))])),
                               ('pipeline-2',
                                Pipeline(steps=[('sex_col',
                                                 SelectCols(cols='sex')),
                                                ('sex_enc',
                                                 OneHotEncoder(sparse_output=False))]))])

Next, run `.fit(X_train)` on your pipe.

In [19]:
pclass_sex_pipe.fit(X_train)

FeatureUnion(transformer_list=[('pipeline-1',
                                Pipeline(steps=[('pclass_col',
                                                 SelectCols(cols='pclass')),
                                                ('pclass_enc',
                                                 OneHotEncoder(sparse_output=False))])),
                               ('pipeline-2',
                                Pipeline(steps=[('sex_col',
                                                 SelectCols(cols='sex')),
                                                ('sex_enc',
                                                 OneHotEncoder(sparse_output=False))]))])

Now combine your `pclass`-`sex` pipeline with a union of another pipeline, this time with `age` and `fare` that are normalized with `StandardScalar`:

In [20]:
from sklearn.preprocessing import StandardScaler

age_fare_pipe = make_union(
    Pipeline([
        ('age_col', SelectCols('age')),
        ('scaler', StandardScaler())
    ]),
    Pipeline([
        ('fare_col', SelectCols('fare')),
        ('scaler', StandardScaler())
    ])
  )
age_fare_pipe

FeatureUnion(transformer_list=[('pipeline-1',
                                Pipeline(steps=[('age_col',
                                                 SelectCols(cols='age')),
                                                ('scaler', StandardScaler())])),
                               ('pipeline-2',
                                Pipeline(steps=[('fare_col',
                                                 SelectCols(cols='fare')),
                                                ('scaler',
                                                 StandardScaler())]))])

In [21]:
pipe = make_union(
      pclass_sex_pipe, age_fare_pipe
  )
pipe

FeatureUnion(transformer_list=[('featureunion-1',
                                FeatureUnion(transformer_list=[('pipeline-1',
                                                                Pipeline(steps=[('pclass_col',
                                                                                 SelectCols(cols='pclass')),
                                                                                ('pclass_enc',
                                                                                 OneHotEncoder(sparse_output=False))])),
                                                               ('pipeline-2',
                                                                Pipeline(steps=[('sex_col',
                                                                                 SelectCols(cols='sex')),
                                                                                ('sex_enc',
                                                                                 OneHotEncoder(sparse_output=False))]))])),
                               ('featureunion-2',
                                FeatureUnion(transformer_list=[('pipeline-1',
                                                                Pipeline(steps=[('age_col',
                                                                                 SelectCols(cols='age')),
                                                                                ('scaler',
                                                                                 StandardScaler())])),
                                                               ('pipeline-2',
                                                                Pipeline(steps=[('fare_col',
                                                                                 SelectCols(cols='fare')),
                                                                                ('scaler',
                                                                                 StandardScaler())]))]))])

Now change `pipe` so that there is a `HistGradientBoostingClassifier` at the end of the pipeline. Then fit your model.

In [22]:
from sklearn.ensemble import HistGradientBoostingClassifier
pipe = make_pipeline(pipe, HistGradientBoostingClassifier())
pipe

Pipeline(steps=[('featureunion',
                 FeatureUnion(transformer_list=[('featureunion-1',
                                                 FeatureUnion(transformer_list=[('pipeline-1',
                                                                                 Pipeline(steps=[('pclass_col',
                                                                                                  SelectCols(cols='pclass')),
                                                                                                 ('pclass_enc',
                                                                                                  OneHotEncoder(sparse_output=False))])),
                                                                                ('pipeline-2',
                                                                                 Pipeline(steps=[('sex_col',
                                                                                                  SelectCols(cols='sex')),
                                                                                                 ('sex_enc',
                                                                                                  OneHotEncoder(sparse_output=False))]))])),
                                                ('featureunion-2',
                                                 FeatureUnion(transformer_list=[('pipeline-1',
                                                                                 Pipeline(steps=[('age_col',
                                                                                                  SelectCols(cols='age')),
                                                                                                 ('scaler',
                                                                                                  StandardScaler())])),
                                                                                ('pipeline-2',
                                                                                 Pipeline(steps=[('fare_col',
                                                                                                  SelectCols(cols='fare')),
                                                                                                 ('scaler',
                                                                                                  StandardScaler())]))]))])),
                ('histgradientboostingclassifier',
                 HistGradientBoostingClassifier())])

Now `.fit()` your pipeline based on your training data. Define that object as `model`.

In [23]:
model = pipe.fit(X_train, y_train)

Use your `X_test` data to make prediction on the test data and name it `y_pred`.

Then use `scikit-learn`'s `classification_report` (import provided) to view the model's performance. Be sure to run it as `print(classification_report ... )` so that it prints to the console.

In [24]:
from sklearn.metrics import classification_report

y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.78      0.89      0.83       144
           1       0.84      0.69      0.76       118

    accuracy                           0.80       262
   macro avg       0.81      0.79      0.80       262
weighted avg       0.81      0.80      0.80       262



Based on your results, does your model have higher precision or recall when Survival = 1?

#The model has higher precision when Survival = 1, but lower recall.

Now take the classification report output and use whatever Large Language Model (ChatGPT, Claude, LLaMa3, whatever). Generate a custom prompt where you provide the LLM the definition of the Titantic dataset and what 1 (Survival) and 0 (Non-Survival) mean. Ask the LLM to now interpret your model's results.

`<!-- Prompt / LLM below -->`
#Please interpret my model's results. I will provide you the model pipeline and the classification report results separately.
Model  pipeline:                                                       
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=42)                                                     from sklearn.pipeline import Pipeline, make_pipeline,FeatureUnion, make_union
from sklearn.preprocessing import OneHotEncoder
from skrub import SelectCols

pclass_sex_pipe = make_union(
    Pipeline([
    ('pclass_col', SelectCols('pclass')),
    ('pclass_enc', OneHotEncoder(sparse_output=False))
  ]),
    Pipeline([
    ('sex_col', SelectCols('sex')),
    ('sex_enc', OneHotEncoder(sparse_output=False))
  ])
)                                                                              pclass_sex_pipe.fit(X_train)                                                                       from sklearn.preprocessing import StandardScaler

age_fare_pipe = make_union(
    Pipeline([
        ('age_col', SelectCols('age')),
        ('scaler', StandardScaler())
    ]),
    Pipeline([
        ('fare_col', SelectCols('fare')),
        ('scaler', StandardScaler())
    ])
  )                                                                             pipe = make_union(
      pclass_sex_pipe, age_fare_pipe
  )                                                                             from sklearn.ensemble import HistGradientBoostingClassifier
pipe = make_pipeline(pipe, HistGradientBoostingClassifier())                               
model = pipe.fit(X_train, y_train)  
#Next I will provide you the classification report results  

from sklearn.metrics import classification_report

y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))                                                             

|           | Precision | Recall | F1-Score | Support |
|-----------|-----------|--------|----------|---------|
| Class 0   |   0.78    |  0.89  |   0.83   |   144   |
| Class 1   |   0.84    |  0.69  |   0.76   |   118   |
| Accuracy  |           |        |   0.80   |   262   |
| Macro Avg |   0.81    |  0.79  |   0.80   |   262   |
| Weighted Avg | 0.81    |  0.80  |   0.80   |   262   |

                                            
#The Titantic dataset was used and 1=(Survival) and 0=(Non-Survival)

`<!-- Show your the LLM's response -->`
#Sure, I'd be happy to help with that. Please go ahead and provide the details of your model pipeline and the classification report results.
#Thank you for providing the model pipeline. Please go ahead and share the classification report results.

#Let's interpret the classification report results of your model on the Titanic dataset:

#Class 0 (Non-Survival):

###Precision: 0.78 (78% of the predicted non-survived instances were actually non-survived).

###Recall: 0.89 (89% of the actual non-survived instances were predicted correctly).

###F1-Score: 0.83 (a good balance between precision and recall).

###Support: 144 instances.

#Class 1 (Survival):

###Precision: 0.84 (84% of the predicted survived instances were actually survived).

###Recall: 0.69 (69% of the actual survived instances were predicted correctly).

###F1-Score: 0.76 (a slightly lower balance between precision and recall compared to Class 0).

###Support: 118 instances.

#Overall:

###Accuracy: 0.80 (80% of the predictions across all classes were correct).

###These results still suggest that the model performs reasonably well, especially in predicting non-survival instances (Class 0), where both precision and recall are relatively high. However, there seems to be room for improvement in predicting survival instances (Class 1), as the recall is lower compared to precision, indicating that the model might be missing some of the actual survived instances.

`<!-- Write in markdown your answer below -->`
#I agree with the LLM's interpretation of the model's performance

# Bonus Problem: Feature Selection

Assume you have the following data for a classification problem.

You've found [`SelectKBest`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html) which will do feature selection based on some metric.

Let's assume you have a big dataset of 200 data points, 10000 features (X variables) and 2 levels in a y variable.



In [ ]:
import numpy as np

n_samples, n_features, n_classes = 200, 10000, 2
rng = np.random.RandomState(42)
X = rng.standard_normal((n_samples, n_features))
y = rng.choice(n_classes, n_samples)

print(X.shape, y.shape)

In [ ]:
print(X[0:10],y[0:10])

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

X_selected = SelectKBest(k=25).fit_transform(X, y)

X_train, X_test, y_train, y_test = train_test_split(
    X_selected, y, random_state=42)
gbc = GradientBoostingClassifier(random_state=42)
gbc.fit(X_train, y_train)

y_pred = gbc.predict(X_test)
accuracy_score(y_test, y_pred)

0.76

## Question

What is a problem with the workflow above?

`<!-- Write in markdown your answer below -->`

Rewrite the workflow to correct the workflow.

#The problem with the workflow above is that we are using the feature selection before splitting the data into train and test sets. Feature selection should be performed with the test data only. We also do not want to use .fit_transform() with the test data either. This method should only be applied to the test set.   

In [38]:
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=42)

selector = SelectKBest(k=25)
X_train_selected = selector.fit_transform(X_train, y_train)

X_test_selected = selector.transform(X_test)

gbc = GradientBoostingClassifier(random_state=42)
gbc.fit(X_train_selected, y_train)

y_pred = gbc.predict(X_test_selected)
accuracy_score(y_test, y_pred)

0.44

#In the corrected workflow, the data is split into the test and train sets first, then feature selection is first fit on the training data. Then the same fitted 'SelectKBest' instance is used to transform the test data.